### ISSUE TITLE
Is `ExtrinsicReport` as Return value for `until` and `until_success` clear enough?

### ISSUE THREAD

**Issue by haerdib** [scs/substrate-api-client#592]
Currently, both functions `watch_until` and `watch_until_sucess` are returning an `ExtrinsicReport`. However, only `until_sucess` returns Events. That only pointed out with comments.
It would be clearer / fail safer if there'd be two Structs.. one returned frum `until` (w.o. Events`) and one with Events, probably a Wrapper struct around the original `ExtrinsicReport`.

See question: https://github.com/scs/substrate-api-client/issues/591#issuecomment-1600543312

**Comment by echevrier** [comment]
Isn't the question more: Is the difference between **`submit_and_watch_extrinsic_until_success`** and **`submit_and_watch_extrinsic_until`** clear enough?

What does **until_success** mean? What do we want to see in the report?

Currently both methods generate an ExtrinsicReport or throw an error when a certain status is reached.  `submit_and_watch_extrinsic_until` allows all status, `submit_and_watch_extrinsic_until_success` only **InBlock** or **Finalized**. Furthermore, only `submit_and_watch_extrinsic_until_success` add the events in the report.
So:
- does success mean an extrinsic is InBlock or Finalized?
- does success mean, we can check the events because the extrinsic succesfully reached the expected status?

Why does `submit_and_watch_extrinsic_until` not fetch the events? If no events were triggered by the extrinsic, then no event are in the report. But if an event has been triggered, then it should be in the report. There is no need to modify the extrinsic report.

**Comment by haerdib** [comment]
So you're suggesting to merge the two functions? E.g. when either InBlock or Finalized is reached, then the Events should be returned? I actually like that idea.. it's way more failsafe. If someone waits for one of these two statuses, one would also like to know if the xt has been successful.

To answer your questions:
- success means that the extrinisc has been successful. A failed extrinsic (e.g. transfer failed because of too little money), may still be "Finalized". But in its failed state.
- Getter calls take extra time and Events take time. In one block there might be a lot of Events, which need to be filtered through to actually get the associated events. So `submit_and_watch_extrinsic_until` is basically the "lightweight" function

**Comment by echevrier** [comment]
So we cannot wait until success. Just hope ;)  
I will let two functions but one report: `submit_and_watch_extrinsic_until `and `submit_and_watch_extrinsic_until_light`and document the difference, i.e. events are not reported for performance reasons. Both functions will accept all status 
If we want to keep a function that only deals with status InBlock and Finalized (user-friendly), currently submit_and_watch_extrinsic_until_success, I will rename this. Maybe `submit_and_watch_extrinsic_until_in_block_or_finalized` ?

**Comment by haerdib** [comment]
Instead of `Light` maybe bare? Because `submit_and_watch_extrinsic_until` (with events) will still need an internal function call for watch. We could simply make that public. 

E.g. change that:

https://github.com/scs/substrate-api-client/blob/a74ea588b222fc0519d29c645c3fdbca6fe90178/src/api/rpc_api/author.rs#L294-L315

to rename `submit_and_watch_until_success` to `submit_and_watch_until` (but with XtStatus input) and `submit_and_watch_until_bare`

What do you think?

Regarding `submit_and_watch_extrinsic_until_in_block_or_finalized` I'm not sure if that's needed.. If a user wants to have simple functions, one could simply define a wrapper function that inserts the "default" status. Not that complicated. So I'm in favour of removing `submit_and_watch_until_success` completely instead of renaming it. An extra function also adds complexity.

**Comment by haerdib** [comment]
Idea:

```rust
async fn submit_and_watch_opaque_extrinsic_until( 
 	&self, 
 	encoded_extrinsic: Bytes, 
 	xt_status: Xtstatus
 ) -> Result<ExtrinsicReport<Self::Hash>> { 
 	let mut report = self 
 		.submit_and_watch_opaque_extrinsic_until_without_events(encoded_extrinsic, xt_status) 
 		.await?; 
        xt_status < in_block {
           return;
        } 
 	let block_hash = report.block_hash.ok_or(Error::BlockHashNotFound)?; 
 	let extrinsic_events = 
 		self.fetch_events_for_extrinsic(block_hash, report.extrinsic_hash).await?; 
 	// Ensure that the extrins has been successful. If not, return an error. 
 	for event in &extrinsic_events { 
 		event.check_if_failed()?; 
 	} 
 	report.events = Some(extrinsic_events); 
 	Ok(report) 
 ```

**Comment by echevrier** [comment]
Our changes are dangerous : `submit_and_watch_extrinsic_until` has a new behaviour. Previously it didn't fetch the events. Now it does so for InBlock and Finalized status. So we have to label the changes we breaking changes to API. 
Is it enough to notify users of breaking changes to the api? 
And if we have breaking changes, shall we keep `submit_and_watch_opaque_extrinsic_until_success` ?

**Comment by haerdib** [comment]
You are right. There's an additional change: it will now fail if the xt has not been successful. 
I could see another way: Introduce a new function (for example `submit_and_watch_extrinsic` that does implement the "new" behaviour, and mark both, `submit_and_watch_opaque_extrinsic_until_success ` and `submit_and_watch_extrinsic_until ` as deprecated and remove them after the one release cycle. What do you think?

**Comment by echevrier** [comment]
`submit_and_watch_extrinsic `is not an option for a new function, it is already implemented: Submit an extrinsic an return a Subscription to watch the extrinsic progress.
In addition, the "new" behaviour is `submit_and_watch_extrinsic_until` but with events.  If we don't want to break the API and stay consistent, the new functionality should be `submit_and_watch_extrinsic_until_and_fetch_events`. We could rename it later, when we remove the deprecated functions.

**Comment by haerdib** [comment]
Right, `submit_and_watch_extrinsic` will not do.

> We could rename it later, when we remove the deprecated functions.

That won't work because marking a function as deprecated indicates that the user should no longer use this function. So renaming after removing does not really make sense, I think.
Also I'd like to avoid too long names. `submit_and_watch_extrinsic_until_and_fetch_events` is really long, especially if there's no `submit_and_watch_extrinsic_until` any more.

I wonder if we're overthinking here. It is a breaking change, but not for the Api. It simply retrieves now events. So yes, it makes it possibly slower, but I guess compared to the time of watching until `InBlock` or `Finalized` it is not that much, I believe.
So maybe we should just do it and document it well?
